### ISSUE TITLE
Ranges [e..e’]

### ISSUE THREAD

**Issue by rolyp** [rolyp/fluid#278]
Add Haskell-style ranges (just for integers). This should also give a nice workflow that we can reuse for #277  and #276. One observation is that we should be able to get test cases for the new features (ranges, list comprehensions and arrays) parsing and evaluating without error, before we start thinking about slicing semantics. This will allow us to proceed in a piecemeal fashion.

- [ ] extend `grammar.ne` to support ranges and regenerate `Parse.ts`
- [ ] add `fluid/example/ranges.fld` test case with some examples
- [ ] add test case to `test/Slice.ts` which simply parses `ranges.fld` (via `openWithImports`)

**Comment by min-nguyen** [comment]
Thanks a lot! Could you briefly give some context behind "Expl", perhaps in comparison to Expr?

**Comment by rolyp** [comment]
Sure. `Expl` are the trace forms (as in `Explanation`, but maybe `Trace` would be better). `ExplValue` is just an `Expl` paired with `Value`.

Roughly speaking there is an `Expl` form for every `Expr` form. They aren’t quite identical because, in particular, `App` traces have a component capturing the execution of the function body.

**Comment by min-nguyen** [comment]
Right, I've pushed my attempt at adding ranges so far - have a look and tell me where I've gone wrong when you get the chance. (Haven't added empty cases for eval_fwd and eval_bwd).

**Comment by rolyp** [comment]
That’s an excellent first attempt. I’ve just pushed a fix; the problem isn’t anything you could have been expected to know about. I think I’ve mentioned that the current implementation does some kind of memoisation (which explains the weird indices kₜ, kᵥ, etc). The memoisation system has an invariant that (in a given execution) the same index k is always assigned the same “content”. You were using the same index kᵥ for all the numbers in the list, which breaks the invariant because the numbers have different values.

To be honest, this memoisation business is mostly unrelated to what we’re currently trying to do, so I will probably take it out. It will simplify things a bit, and it will only take a couple of hours to do. I’ll do that either late afternoon or tomorrow morning.

In the meantime, your test case is now failing because of the missing `eval_fwd` and `eval_bwd` cases, which I guess is the point you were hoping to get to.

Good work!

**Comment by rolyp** [comment]
Btw I’m busy most of the day now but I’ll be free from 4ish this afternoon.

**Comment by min-nguyen** [comment]
Thanks! I've been trying to figure out eval_fwd and eval_bwd (and the whole project) for a while, but I doubt I'll be able to get anywhere further until we get to talk on Friday.

**Comment by rolyp** [comment]
You can probably forget `eval_fwd` and `eval_bwd` at least until list comprehensions are finished. Dummy (empty) cases for those functions should allow the tests to pass. If they do then you can create a pull request for this branch and we can consider this initial step complete.

**Comment by min-nguyen** [comment]
The tests aren't passing when I'm using empty cases unfortunately! Assertion failures.

**Comment by rolyp** [comment]
Ok if you push your latest changes I’ll take a look.

**Comment by min-nguyen** [comment]
Cheers, have done so.

**Comment by rolyp** [comment]
Cool. You’ll also need to do something to turn that into meaningful program, e.g. by concatenation (currently it applies a list to a list, which doesn’t work!).

Unfortunately there is still a problem: ranges introduce a new kind of “terminal” form into the language (by “terminal” form I mean an expression that directly produces a value, such a constructor or constant, rather than one which delegates to another expression to do the work, such as a function call or pattern match). This breaks an assumption that exists in at least one place (`Expl.explChild`), namely that every data value (such as a list) is made by a data constructor (such a list constructor). We’ll also run into this issue with list comprehensions, which also provide another way of making lists.

There are a couple of ways we could address this. Ultimately what we want is to introduce a desugaring phase, and define ranges (and comprehensions) as syntactic sugar. The main challenge there is that sugaring needs to be “invertible” (via a trace form for sugaring) so that we can do backward slicing and put information back to the original program, not the desugared programs. That’s probably the way we want to go in the longer term, especially as the language acquires more realistic features. But it is new to the slicing setting and will require some work to get right, so probably beyond what we want to do now.

There are alternatives we could look at in the meantime, so let’s talk about this tomorrow.
